libs/utils/include/utils/vector.h

```diff
@@ -55,6 +55,214 @@ static inline bool insert_sorted_unique(std::vector<T>& v, T item) {
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+
+    vector() noexcept = default;
+
+    explicit vector(size_type count) noexcept : TrivialVectorBase(sizeof(T), count) { }
+
+    explicit vector(size_type count, const value_type& proto) noexcept
+            : TrivialVectorBase(sizeof(T), count) {
+        std::generate_n(begin(), count, [proto]{ return proto; });
+    }
+
+    // --------------------------------------------------------------------------------------------
+
+    iterator begin() noexcept { return reinterpret_cast<iterator>(TrivialVectorBase::begin()); }
+
+    iterator end() noexcept { return reinterpret_cast<iterator>(TrivialVectorBase::end(sizeof(T))); }
+
+    const_iterator begin() const noexcept { return reinterpret_cast<const_iterator>(TrivialVectorBase::begin()); }
+
+    const_iterator end() const noexcept { return reinterpret_cast<const_iterator>(TrivialVectorBase::end(sizeof(T))); }
+
+    reverse_iterator rbegin() noexcept { return reverse_iterator(end()); }
+
+    const_reverse_iterator rbegin() const noexcept { return const_reverse_iterator(end()); }
+
+    reverse_iterator rend() noexcept { return reverse_iterator(begin()); }
+
+    const_reverse_iterator rend() const noexcept { return const_reverse_iterator(begin()); }
+
+    const_iterator cbegin() const noexcept { return begin(); }
+
+    const_iterator cend() const noexcept { return end(); }
+
+    const_reverse_iterator crbegin() const noexcept { return rbegin(); }
+
+    const_reverse_iterator crend() const noexcept { return rend(); }
+
+    // --------------------------------------------------------------------------------------------
+
+    size_type size() const noexcept { return mItemCount; }
+
+    size_type capacity() const noexcept { return mCapacity; }
+
+    bool empty() const noexcept { return mItemCount == 0; }
+
+    // --------------------------------------------------------------------------------------------
+
+    reference operator[](size_type n) noexcept { return *(begin() + n); }
+
+    const_reference operator[](size_type n) const noexcept { return *(begin() + n); }
+
+    reference front() noexcept { return *begin(); }
+
+    const_reference front() const noexcept { return *begin(); }
+
+    reference back() noexcept { return *(end() - 1); }
+
+    const_reference back() const noexcept { return *(end() - 1); }
+
+    value_type* data() noexcept { return begin(); }
+
+    const value_type* data() const noexcept { return begin(); }
+
+    // --------------------------------------------------------------------------------------------
+
+    void push_back(const_reference v) {
+        assert_capacity(size() + 1, sizeof(T));
+        *end() = v;
+        mItemCount++;
+    }
+
+    template<typename ... ARGS>
+    reference emplace_back(ARGS&& ... args) {
+        assert_capacity(size() + 1, sizeof(T));
+        T::T(end())(std::forward<ARGS>(args)...);
```

Comment:
I'm probably missing something obvious, but I can't parse this line. What's `T::T` suppose to be? When I try to create a `struct Foo {}; utils::vector<Foo>`, I get an error.

Reply:
It's a `T::T(…)`, a constructor.

Reply:
This is what I mean:

```
struct Foo { };
utils::vector<Foo> v;
v.emplace_back();
```

```
../../libs/utils/include/utils/vector.h:209:12: error: no member named 'T' in 'Foo'
        T::T(end())(std::forward<ARGS>(args)...);
```

Reply:
Yeah I completely goofed up there.
The code should have been `new T(end())(...)`.  Because it's a template it didn't fail at compile time until I actually tried to use it :-)

Reply:
I was thinking this was some bizarre new C++ syntax I hadn't seen before, which wouldn't have surprised me :)